src/transaction/confirm.rs

```diff
@@ -0,0 +1,601 @@
+        }
+    }
+}
+
+impl Default for ConfirmParams {
+    fn default() -> Self {
+        ConfirmParams::mined()
+    }
+}
+
+/// A future that resolves once a transaction is confirmed.
+#[derive(Debug)]
+pub struct ConfirmFuture<T: Transport> {
+    web3: Web3Unpin<T>,
+    /// The transaction hash that is being confirmed.
+    tx: H256,
+    /// The confirmation parameters (like number of confirming blocks to wait
+    /// for and polling interval).
+    params: ConfirmParams,
+    /// The current block number when confirmation started. This is used for
+    /// timeouts.
+    starting_block_num: Option<U256>,
+    /// The current state of the confirmation.
+    state: ConfirmState<T>,
+}
+
+/// The state of the confirmation future.
+enum ConfirmState<T: Transport> {
+    /// The future is in the state where it needs to setup the checking future
+    /// to see if the confirmation is complete. This is used as a intermediate
+    /// state that doesn't actually wait for anything and immediately proceeds
+    /// to the `Checking` state.
+    Check,
+    /// The future is waiting for the block number and transaction receipt to
+    /// make sure that enough blocks have passed since the transaction was
+    /// mined. Note that the transaction receipt is retrieved everytime in case
+    /// of ommered blocks.
+    Checking(CheckFuture<T>),
+    /// The future is waiting for the block filter to be created so that it can
+    /// wait for blocks to go by.
+    CreatingFilter(CompatCreateFilter<T, H256>, U256, u64),
+    /// The future is waiting for new blocks to be mined and added to the chain
+    /// so that the transaction can be confirmed the desired number of blocks.
+    WaitingForBlocks(CompatFilterFuture<T, H256>),
+    /// The future is waiting for a poll timeout. This state happens when the
+    /// node does not support block filters for the given transport (like Infura
+    /// over HTTPS) so we need to fallback to polling.
+    PollDelay(Delay, U256),
+    /// The future is checking that the current block number has reached a
+    /// certain target after waiting the poll delay.
+    PollCheckingBlockNumber(CompatCallFuture<T, U256>, U256),
+}
+
+impl<T: Transport> ConfirmFuture<T> {
+    /// Create a new `ConfirmFuture` with a `web3` provider for the specified
+    /// transaction hash and with the specified parameters.
+    pub fn new(web3: &Web3<T>, tx: H256, params: ConfirmParams) -> Self {
+        ConfirmFuture {
+            web3: web3.clone().into(),
+            tx,
+            params,
+            starting_block_num: None,
+            state: ConfirmState::Check,
+        }
+    }
+}
+
+impl<T: Transport> Future for ConfirmFuture<T> {
+    type Output = Result<TransactionReceipt, ExecutionError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let unpinned = self.get_mut();
+        loop {
+            unpinned.state = match &mut unpinned.state {
```

Comment:
nice state machine. Is this a common pattern in rust? I was wondering if it made sense for the SendFuture (as the flow is only linear) but here it makes a lot of sense with all the jumps.

Reply:
It is a common pattern for futures and is how the generated async code by the compiler looks like in theory.